sdks/java/core/src/main/java/org/apache/beam/sdk/io/fs/ResourceIdTester.java

```diff
@@ -77,31 +80,58 @@ private static void validateResolvingIds(
     allResourceIds.add(dir2);
 
     // ResourceIds in equality groups.
-    new EqualsTester()
-        .addEqualityGroup(file1)
-        .addEqualityGroup(file2, file2a)
-        .addEqualityGroup(dir1, dir1.getCurrentDirectory())
-        .addEqualityGroup(dir2, dir2a, dir2.getCurrentDirectory())
-        .addEqualityGroup(baseDirectory, file1.getCurrentDirectory(), file2.getCurrentDirectory())
-        .testEquals();
+    assertEqualityGroups(
+        Arrays.asList(
+            Arrays.asList(file1),
+            Arrays.asList(file2, file2a),
+            Arrays.asList(dir1, dir1.getCurrentDirectory()),
+            Arrays.asList(dir2, dir2a, dir2.getCurrentDirectory()),
+            Arrays.asList(
+                baseDirectory, file1.getCurrentDirectory(), file2.getCurrentDirectory())));
 
     // ResourceId toString() in equality groups.
-    new EqualsTester()
-        .addEqualityGroup(file1.toString())
-        .addEqualityGroup(file2.toString(), file2a.toString())
-        .addEqualityGroup(dir1.toString(), dir1.getCurrentDirectory().toString())
-        .addEqualityGroup(dir2.toString(), dir2a.toString(), dir2.getCurrentDirectory().toString())
-        .addEqualityGroup(
-            baseDirectory.toString(),
-            file1.getCurrentDirectory().toString(),
-            file2.getCurrentDirectory().toString())
-        .testEquals();
+    assertEqualityGroups(
+        Arrays.asList(
+            Arrays.asList(file1.toString()),
+            Arrays.asList(file2.toString(), file2a.toString()),
+            Arrays.asList(dir1.toString(), dir1.getCurrentDirectory().toString()),
+            Arrays.asList(dir2.toString(), dir2a.toString(), dir2.getCurrentDirectory().toString()),
+            Arrays.asList(
+                baseDirectory.toString(),
+                file1.getCurrentDirectory().toString(),
+                file2.getCurrentDirectory().toString())));
 
     // TODO: test resolving strings that need to be escaped.
     //   Possible spec: https://tools.ietf.org/html/rfc3986#section-2
     //   May need options to be filesystem-independent, e.g., if filesystems ban certain chars.
   }
 
+  /**
+   * Asserts that all elements in each group are equal to each other but not equal to any other
+   * element in another group.
+   */
+  private static <T> void assertEqualityGroups(List<List<T>> equalityGroups) {
+    for (int i = 0; i < equalityGroups.size(); ++i) {
+      List<T> current = equalityGroups.get(i);
+      for (int j = 0; j < current.size(); ++j) {
+        for (int k = 0; k < current.size(); ++k) {
```

Comment:
Could you do `int k = j` instead here to avoid duplicate comparisons?

Reply:
Since equals is implemented within the class a.equals(b) is not the same thing as b.equals(a) if a and b are two different classes, we also test a.equals(a) for posterity.

Reply:
Ah, I didn't consider that case. Ok that makes sense now.